lib/usd/translators/meshWriter.cpp

```diff
@@ -54,6 +59,57 @@
 #include <string>
 #include <vector>
 
+MObject mayaFindUpstreamOrigMesh(const MObject& mesh)
+{
+    MStatus stat;
+
+    // NOTE: (yliangsiew) If there's a skinCluster, find that first since that
+    // will be the intermediate to the blendShape node. If not, just search for any
+    // blendshape deformers upstream of the mesh.
+    MObject searchObject;
+    MObject skinCluster;
+    stat = mayaGetSkinClusterConnectedToMesh(mesh, skinCluster);
+    if (stat) {
+        searchObject = MObject(skinCluster);
+    } else {
+        searchObject = MObject(mesh);
+    }
+
+    // TODO: (yliangsiew) Problem: if there are _intermediate deformers between blendshapes,
+    // then what do we do? Like blendshape1 -> wrap -> blendshape2. This won't find
+    // that correctly...
```

Comment:
Maya's GPU Deformer has a similar problem where it has to parse deformation graphs and figure out what can be supported by GPU evaluation and what is not. We do something similar to your itDg except we don't set the node type filter and do that part manually in the loop, testing to see if each upstream node is something supported as we go.

You could do something similar here. The code knows the set of supported deformer types that can be exported to UsdSkel so you can search upwards and find either the static base mesh, or the set of boundary nodes which are animated but not supported.

Reply:
Thinking about it more carefully I think there is an issue with the MItDependencyGraph usage you have. The intent of the code is to follow the deformation chain upstream to find the original mesh. That means you should only be following connections which are carrying mesh data, not other data.

Imagine a scene where there are two separate meshes being driven by blend shapes. If the weight attribute of blendShape1 is driven by the weight attribute of blendShape2 then when searching for the upstream original mesh of blendShape1 the iterator will find both blendShape1 and blendShape2. If blendShape2 is found first then an incorrect inputGeo (originalMesh2) can be returned, a mesh which can have no relationship to deformedMesh1.

Our GPU Override code solves this by limiting the search plugs to only geometry plugs which affect the output geometry & which must match the output geometry's topology. So for a blend shape the upstream search plugs are inputTargetGeom and inputGeom. We have lots of cases where inputTargetGeom has incoming deformation.

Reply:
Could you elaborate a little on this setup that you're referring to? I've tried to re-create as described above. In the case shown in the diagram, whether starting from pCube1baseShape or pCube2baseShape, I think either should find the correct corresponding origShape node (since I exit iteration once the mesh is found, thus it shouldn't find the other blendshape node, regardless of the fact that there is a connection between the two blendshape nodes by virtue of the weight attribute of one shape driving the other). I might be misunderstanding here.

![image](https://user-images.githubusercontent.com/74624210/100806241-f6234a00-33e4-11eb-9d22-c574e05a6730.png)

Reply:
You are correct the code you have can't return the input of the wrong blend shape. I misunderstood the power of MFnGeometryFilter::inputShapeAtIndex().

The problem with this function is if there are more deformers upstream of the blend shape's input geometry then inputShapeAtIndex skips over them and goes directly to a shape. There could be other deformers between the discovered mesh and the input to the blend shape node. For example those tweak nodes do not have to be empty, or there could be other animated deformers present. In that case the original mesh data you read won't match the mesh data Maya has at the input of the blend shape, and the deformation exported to USD won't match Maya.

In addition to that the discovered mesh node doesn't have to be static. The mesh can have an input mesh connection that can be animated with deformation as well.

The problem is Maya can support any combination of these nodes together. The code here must do more pattern matching to ensure that the deformation described by the DG is Maya matches a case that the exporter understands. 

My suggestion to you would be to be extremely specific about the configuration of supported nodes to limit what is allowed to be exported and what is not. Trying to handle exporting arbitrary graphs of deformers is a huge task that I don't want this change to get bogged down in. Focus this change on doing a really good job exporting all the features of a blend shape node (which you already have) but in a simple graph context. The only thing missing is verifying that the graph of deformer nodes matches something the exporter can handle.

Reply:
I have added a new `ignoreWarnings` flag to the `mayaUSDExport` command, which
will control this behaviour (defaulting to `false`). If any unrecognized node is
found, the export will be blocked. Currently the only nodes whitelisted are any
nodes supporting the Fn sets:

- kGroupParts
- kMesh (i.e. mesh piping output to another mesh)

For tweak nodes, they are checked to see if they have any active effect; if not,
they are whitelisted as well. For deformers, they are checked to see if their
envelopes are set to `0.0f` and if so, whitelisted as well.

All other types of nodes will trigger an error message on export. Hopefully this
addresses the issue that you're bringing up?

Reply:
Yes my concerns have been addressed. You've done a good job validating that the deformation expressed in the DG matches the capabilities of the export code.

There is one more thing to test for mesh nodes. Meshes can have tweaks embedded in them in the pnts attribute. I think normally it'll be empty but you can put something in there through MFnMesh::setPoint, in which case the mesh won't be a pass through and safe for your whitelist.

Reply:
> There is one more thing to test for mesh nodes. Meshes can have tweaks embedded in them in the pnts attribute. I think normally it'll be empty but you can put something in there through MFnMesh::setPoint, in which case the mesh won't be a pass through and safe for your whitelist.

This should now be addressed as part of `7c733f5`.